src/sql/src/query_model/hir.rs

```diff
@@ -69,6 +73,18 @@ impl FromHir {
                 }
                 box_id
             }
+            HirRelationExpr::Filter { input, predicates } => {
+                let input_box = self.generate_internal(input);
+                // Wrap the input within a select so the referenced columns are
+                // in the expected order
+                let select_id = self.wrap_within_select(input_box);
+                for predicate in predicates {
+                    let expr = self.generate_expr(predicate, select_id);
+                    self.add_predicate(select_id, expr);
+                }
+                select_id
+            }
+
             HirRelationExpr::Project { input, outputs } => {
                 let input_box_id = self.generate_internal(input);
```

Comment:
The recursive call structure for `HirRelationExpr::Map` and `HirRelationExpr::Project` is different.

The `Map` case uses `generate_select` and piggy-backs on the resulting box, whereas `Project` uses `generate_internal` (as the remaining cases) and explicitly creates a `Select` box for the current case with `make_select_box`. Can we align the two (preferrably make `Map` like `Project`, to be consistent with the remaining cases)?

Reply:
`Map` appends new column to the input relation whereas `Project` re-shuffles the projection of the input relation. Re-using the input box when lowering a `Project` operator feels like applying a transformation in place, making the output of the traceability of this process harder.

Also, when adding support for self-referencing `Map`s we will actually add a new Select box for every scalar expression in the `Map`. Making it consistent with `Project`.